ansible/roles/jenkins-worker/templates/rpi_docker-node-exec.sh.j2

```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
+
+set -e
+
+OPTIND=1
+mode=js
+flaky=run
+subsets=
+run=
+version=
+
+while getopts "afs:r:v:" opt; do
+    case "$opt" in
+    a)
+        mode=addons
+        ;;
+    f)
+        flaky=dontcare
+        ;;
+    s)
+        if [[ "$OPTARG" =~ ^[0-9]+$ ]]; then
+          subsets=$OPTARG
+        else
+          echo "Bad -s value"
+          exit 1
+        fi
+        ;;
+    r)
+        if [[ "$OPTARG" =~ ^[0-9]+$ ]]; then
+          run=$OPTARG
+        else
+          echo "Bad -r value"
+          exit 1
+        fi
+        ;;
+    v)
+        if [[ "$OPTARG" =~ ^[a-z]+$ ]]; then
+          version=$OPTARG
+        else
+          echo "Bad -v value"
+          exit 1
+        fi
+        ;;
+    esac
+done
+
+if test "$version" = ""; then
+  echo "Did not provide the Raspbian version [-v]"
+  exit 1
+fi
+
+if test "$mode" != "addons"; then
+  if test "$run" = ""; then
+    echo "Did not provide the subset number of this run [-r]"
+    exit 1
+  fi
+
+  if test "$subsets" = ""; then
+    echo "Did not provide number of subsets [-s]"
+    exit 1
+  fi
+fi
+
+echo "Running on $version with mode=$mode, run=$run/$subsets and flaky=$flaky"
+dir=$(pwd)
+
+if test "$mode" = "addons"; then
+  set -x
+  docker exec node-ci-$version /bin/sh -c "cd $dir && PYTHON=python FLAKY_TESTS=$flaky make test-ci-native"
+else
+  set -x
+  docker exec node-ci-$version /bin/sh -c "cd $dir && CC=should-not-compile CXX=should-not-compile PYTHON=python FLAKY_TESTS=$flaky TEST_CI_ARGS=--run=${run},${subsets} make test-ci-js"
```

Comment:
I would prefer the make command to remain in the Jenkins script. Could this be something like
```
  docker exec node-ci-$version /bin/sh -c "cd $dir && $1"
```
so that Jenkins could do
```
sudo docker-node-exec.sh -v $raspbian "PYTHON=python FLAKY_TESTS=$FLAKY_TESTS_MODE make test-ci-native"
```
?
It's better to have all of the logic to run the tests in the same place. However, if this is not simple to do or there's a good reason that I'm not thinking of, please go ahead as is. We can improve later.

Reply:
@joaocgreis yeah but the problem is that I don't really want to allow unsanitised input to be run through that script. `$1` could potentially be used to break out of the script. I'm open to suggestions on how to do that, just be aware that that script is being run as `root` and in addition, having unbounded access to `docker` essentially gives you the same level of access. If we're comfortable that any `$*` can be safely passed on then that's fine, that's all I was trying to achieve by putting logic in the script. (also be aware that if you have access to edit this stuff in Jenkins then you also have access to login to these Pi's as user `pi` which has full `sudo` access anyway, including the ability to edit this script to make changes into the future if that's what you're worried about).

Reply:
Indeed we should be careful here. What about passing it in a text file? Script:
```
  docker exec node-ci-$version /bin/sh -c "cd $dir && source out/test-ci"
```
And Jenkins:
```
[ -e out/test-ci ] && exit 1   # Make sure it doesn't exit
echo "PYTHON=python FLAKY_TESTS=$FLAKY_TESTS_MODE make test-ci-native" > out/test-ci
sudo docker-node-exec.sh -v $raspbian
```
Doesn't need to be `out/test-ci`, this is just taking advantage that `out/` is already there from the binaries.

Reply:
oh, that's a clever idea, then it's better constrained, I'll give that a try